### PR TITLE
[python package] added get_ref_chain function to Dataset object, used it in train() to…

### LINKED ISSUE #714: [python package]: datasets with common dataset.reference
When doing lgb.train with a training and validation set, the train function sets the reference of the valid_data to be train_set (line 131 engine.py).  this set_reference then checks if the reference of the valid_data is the train_set (line 1009 basics.py), and if not, sets the reference of valid_data to be train_set.

This throws an error if the train_set was constructed with free_raw_data=True.  In particular, this is an issue if I want to:
1. create a single big lgb.Dataset object with all the data
2. free the raw data used
3. create subsets of the single big lgb.Dataset for training and validation

It seems like this should work, given that both train and validation have the same bin structure, as they are subsets of a single lgb.Dataset object.

The problem seems like it could be solved by changing the set reference function (line 1009 basics.py), so that instead of checking if self.reference is reference, it instead  iterates up the entire chain of potentially chained references and detects if the ultimate parent of both self.reference and reference are the same. 

Example code for reproducing error:
~~~
import lightgbm as lgb
import numpy as np
import pandas as pd
import sys
import platform
print(sys.version)
# 3.6.0 |Anaconda 4.3.0 (64-bit)| (default, Dec 23 2016, 11:57:41) [MSC v.1900 64 bit (AMD64)]
print(platform.system(), platform.release(), platform.version())
# Windows 2012ServerR2 6.3.9600

# 1. generate some data
n = 1000
tmp_df = pd.DataFrame({'x1':np.random.normal(size=n),'x2':np.random.normal(size=n), 'y':np.random.normal(size=n) })
tmp_df.head()
tmp_dat = lgb.Dataset(tmp_df[['x1', 'x2']], tmp_df['y'])
# generate training/val subsets
tmp_dat_train = tmp_dat.subset(np.arange(int(n*.8)))
tmp_dat_val   = tmp_dat.subset(np.arange(int(n*.8), n))

#2. print the various references of the various lgb.Dataset objects
print(tmp_dat, tmp_dat.reference)
print(tmp_dat_train, tmp_dat_train.reference)
print(tmp_dat_val, tmp_dat_val.reference)
# <lightgbm.basic.Dataset object at 0x000000AF75B196A0> None
# <lightgbm.basic.Dataset object at 0x000000AF75B196D8> <lightgbm.basic.Dataset object at 0x000000AF75B196A0>
# <lightgbm.basic.Dataset object at 0x000000AF75B19E10> <lightgbm.basic.Dataset object at 0x000000AF75B196A0>
# # note: tmp_dat_val and tmp_dat_train share the same reference, which is tmp_dat

# 3. do training, where the validation's reference is the training
params = {'objective':'regression_l2',
         'metric':'rmse'}
gbm = lgb.train(params, tmp_dat, num_boost_round=20, valid_sets=[tmp_dat, tmp_dat_val])
# everything runs, no errors

#4. do training, where the validation's reference is NOT the training, but
#    rather validation and training share the same reference
params = {'objective':'regression_l2',
         'metric':'rmse'}
gbm = lgb.train(params, tmp_dat_train, num_boost_round=20, valid_sets=[tmp_dat_train, tmp_dat_val])
# LightGBMError: Cannot set reference after freed raw data, set free_raw_data=False when construct Dataset to avoid this.
~~~

### PR DESCRIPTION
**Pull request by j-mark-hou**
… compare if training and validation have common reference ancestors.

Basically, in engine.py, the train() function no longer directly sets the reference of valid_data to be train_set, but rather checks to see if valid_data and train_set are both derived from the same 'ancestor', and only does does the reference setting if not.   In order to do this check I implemented a method for the Dataset object which returns the dataset itself, its reference, the reference of its reference, etc.

This should resolve #714 

I built lightgbm with the changes and ran the example code in #714 which showed an error, and with the modifications there were no more errors.  Let me know if there are other checks I should do.

### CODE REVIEW

**Comment by guolinke** [comment]
@j-mark-hou 
you solution is to avoid calling ```valid_data.set_reference(train_set)``` . 
I think it cannot solve the issue. 
I think a better solution is checking the dataset have common reference or not here (https://github.com/Microsoft/LightGBM/blob/master/python-package/lightgbm/basic.py#L1009-L1013) .

**Comment by j-mark-hou** [comment]
Sure, I can go ahead and make that change.  My initial impression was that a function named set_reference should actually force a setting of the reference irrespective of whether or not they already share the same ancestor, but if that's the only way to go I'm happy to move the check in there.

File: python-package/lightgbm/basic.py [hunk 1]
```diff
@@ -1174,6 +1175,25 @@ def num_feature(self):
...
+
+        Returns
+        -------
+        chain of references of self : set of Dataset objects
+        '''
+        head = self
+        ref_chain=[]
+        while len(ref_chain)<ref_limit:
+            ref_chain += ([head] if isinstance(head, Dataset) else [])
+            if isinstance(head, Dataset) and (head.reference is not None) and (head.reference not in ref_chain):
```
Comment (guolinke) [inline python-package/lightgbm/basic.py]: I think the reference with handle=None should be excluded as well

**Comment by guolinke** [comment]
can you fix the pep8 format checking?

File: python-package/lightgbm/basic.py (hunk 1 above)
Comment (j-mark-hou) [inline python-package/lightgbm/basic.py]: Something like this? 


```
head = self
ref_chain=[]
while len(ref_chain)<ref_limit:
    if isinstance(head, Dataset) and head.handle is not None:
        ref_chain += ([head] if isinstance(head, Dataset) else [])
        if (head.reference is not None) and (head.reference not in ref_chain):
            head = head.reference
        else:
            break
    else:
        break
return(set(ref_chain))
```

Comment (guolinke) [inline python-package/lightgbm/basic.py]: yeah

Comment (j-mark-hou) [inline python-package/lightgbm/basic.py]: So I made this change and re-built but excluding things with handle = None breaks the example in #714.  In particular, all of tmp_dat, tmp_dat_train, tmp_dat_val have handle=None.  To be totally honest I'm not super clear on what handle is.  Is there an easy way around this?

Comment (guolinke) [inline python-package/lightgbm/basic.py]: sorry, checking the handle=None is not need. You can revert this change.

File: python-package/lightgbm/basic.py [hunk 2]
```diff
@@ -1004,7 +1004,8 @@ def set_reference(self, reference):
         self.set_categorical_feature(reference.categorical_feature)
         self.set_feature_name(reference.feature_name)
         self._set_predictor(reference._predictor)
-        if self.reference is reference:
+        # we're done if self and reference share a common upstrem reference
+        if self.get_ref_chain().intersection(reference.get_ref_chain()):
             return
```
Comment (guolinke) [inline python-package/lightgbm/basic.py]: should we add ```self.reference = reference``` before ```return``` ?

Comment (j-mark-hou) [inline python-package/lightgbm/basic.py]: my impression is that in this situation, which self.reference and reference share a common ancestor, we would could just not bother with setting `self-reference=reference'.  otherwise we get the same issue that motivated this pr.  perhaps I'm misunderstanding you?

File: python-package/lightgbm/basic.py [hunk 3]
```diff
@@ -1174,6 +1175,28 @@ def num_feature(self):
...
+        '''
+        Gets a chain of Dataset objects, starting with r, then going to r.reference if exists,
+            then to r.reference.reference, etc. until we hit ref_limit or a reference loop
+
+        Returns
+        -------
+        chain of references of self : set of Dataset objects
+        '''
+        head = self
+        ref_chain = []
```
Comment (wxchan) [inline python-package/lightgbm/basic.py]: why not init it as set()?

Comment (j-mark-hou) [inline python-package/lightgbm/basic.py]: yeah, that's true.  Should I just open up another PR to fix this?

Comment (wxchan) [inline python-package/lightgbm/basic.py]: @j-mark-hou you can open another one. Can you also add a simple test case for your issue in test folder (or modify an existed one maybe)?

Comment (j-mark-hou) [inline python-package/lightgbm/basic.py]: yeah, sounds good.  would a test case based on the one point 4 in #714 work?

Comment (wxchan) [inline python-package/lightgbm/basic.py]: Yes, the test case may fail before, and fixed by this pr.
